ophyd/positioner.py

```diff
@@ -85,67 +81,30 @@ def move(self, position, wait=True, moved_cb=None, timeout=30.0):
         self._reset_sub(self._SUB_REQ_DONE)
 
-        is_subclass = (self.__class__ is not Positioner)
-        if not is_subclass:
-            # When not subclassed, Positioner acts as a soft positioner,
-            # immediately 'moving' to the target position when requested.
-            self._started_moving = True
-            self._moving = False
+        status = MoveStatus(self, position, timeout=timeout)
 
-        status = MoveStatus(self, position)
-        if wait:
-            t0 = time.time()
-
-            def check_timeout():
-                return timeout is not None and (time.time() - t0) > timeout
-
-            while not self._started_moving:
-                time.sleep(0.05)
-
-                if check_timeout():
-                    raise TimeoutError('Failed to move %s to %s '
-                                       'in %s s (no motion)' %
-                                       (self.name, position, timeout))
-
-            while self.moving:
-                time.sleep(0.05)
+        if moved_cb is not None:
+            status.finished_cb = partial(moved_cb, obj=self)
+            # the status object will run this callback when finished
 
-                if check_timeout():
-                    raise TimeoutError('Failed to move %s to %s in %s s' %
-                                       (self.name, position, timeout))
-
-            status._finished()
-
-        else:
-            if moved_cb is not None:
-                self.subscribe(moved_cb, event_type=self._SUB_REQ_DONE,
-                               run=False)
-
-            self.subscribe(status._finished,
-                           event_type=self._SUB_REQ_DONE, run=False)
-
-        if not is_subclass:
-            self._set_position(position)
-            self._done_moving()
+        self.subscribe(status._finished, event_type=self._SUB_REQ_DONE,
+                       run=False)
 
         return status
 
-    def _done_moving(self, timestamp=None, value=None, **kwargs):
+    def _done_moving(self, success=True, timestamp=None, value=None, **kwargs):
         '''Call when motion has completed.  Runs SUB_DONE subscription.'''
+        if success:
+            self._run_subs(sub_type=self.SUB_DONE, timestamp=timestamp,
+                           value=value)
 
-        self._run_subs(sub_type=self.SUB_DONE, timestamp=timestamp,
-                       value=value, **kwargs)
-
-        self._run_subs(sub_type=self._SUB_REQ_DONE, timestamp=timestamp,
-                       value=value, success=True,
-                       **kwargs)
+        self._run_subs(sub_type=self._SUB_REQ_DONE, success=success,
+                       timestamp=timestamp)
         self._reset_sub(self._SUB_REQ_DONE)
 
     def stop(self):
         '''Stops motion'''
-
-        self._run_subs(sub_type=self._SUB_REQ_DONE, success=False)
-        self._reset_sub(self._SUB_REQ_DONE)
+        self._done_moving(success=False)
```

Comment:
Ah, great, this solves an issue I was having with the undulator motors last night.